src/main/java/org/broadinstitute/hellbender/tools/variantdb/ArrayExtractCohortEngine.java

```diff
@@ -0,0 +1,363 @@
+        }
+        final StorageAPIAvroReader storageAPIAvroReader = new StorageAPIAvroReader(cohortTableRef);
+        createVariantsFromUngroupedTableResult(storageAPIAvroReader);
+    }
+
+
+    private void createVariantsFromUngroupedTableResult(final GATKAvroReader avroReader) {
+
+        // stream out the data and sort locally
+        final org.apache.avro.Schema schema = avroReader.getSchema();
+        final Set<String> columnNames = new HashSet<>();
+        schema.getFields().forEach(field -> columnNames.add(field.name()));
+
+        Comparator<GenericRecord> comparator = this.useCompressedData ? COMPRESSED_PROBE_ID_COMPARATOR : UNCOMPRESSED_PROBE_ID_COMPARATOR;
+        SortingCollection<GenericRecord> sortingCollection =  getAvroProbeIdSortingCollection(schema, localSortMaxRecordsInRam, comparator);
+        for ( final GenericRecord queryRow : avroReader ) {
+            sortingCollection.add(queryRow);
+        }
+
+        sortingCollection.printTempFileStats();
+
+        // iterate through records and process them
+        final List<GenericRecord> currentPositionRecords = new ArrayList<>(sampleIdMap.size() * 2);
+        long currentProbeId = -1;
+
+        for ( final GenericRecord sortedRow : sortingCollection ) {
+            long probeId;
+            if (useCompressedData) {
+                final long rawData = (Long) sortedRow.get(SchemaUtils.RAW_ARRAY_DATA_FIELD_NAME);
+                RawArrayData data = RawArrayData.decode(rawData);
+                probeId = data.probeId;
+            } else {
+                probeId = (Long) sortedRow.get("probe_id");
+            }
+
+            if ( probeId != currentProbeId && currentProbeId != -1 ) {
+                ++totalNumberOfSites;
+                processSampleRecordsForLocation(currentProbeId, currentPositionRecords, columnNames);
+                currentPositionRecords.clear();
+            }
+
+            currentPositionRecords.add(sortedRow);
+            currentProbeId = probeId;
+        }
+
+        if ( ! currentPositionRecords.isEmpty() ) {
+            ++totalNumberOfSites;
+            processSampleRecordsForLocation(currentProbeId, currentPositionRecords, columnNames);
+        }
+    }
+
+    private void processSampleRecordsForLocation(final long probeId, final Iterable<GenericRecord> sampleRecordsAtPosition, final Set<String> columnNames) {
+        final List<VariantContext> unmergedCalls = new ArrayList<>();
+        final Set<String> currentPositionSamplesSeen = new HashSet<>();
+        boolean currentPositionHasVariant = false;
+
+        final ProbeInfo probeInfo = probeIdMap.get(probeId);
+        if (probeInfo == null) {
+            throw new RuntimeException("Unable to find probeInfo for " + probeId);
+        }
+
+        final String contig = probeInfo.contig;
+        final long position = probeInfo.position;
+        final Allele refAllele = Allele.create(refSource.queryAndPrefetch(contig, position, position).getBaseString(), true);
```

Comment:
i wonder if it would be easier to get the ref allele from the probe_info table - since we already use it. and then we might not need the reference as an input?

Reply:
Something to think about, although we still need the reference when we initialize the vcf writer to get the sequence dictionary for the header